src/expint.jl

```diff
@@ -0,0 +1,410 @@
+    
+    for k = 0:100
+        a_pre = Δ_prod_fact + a*Δ*(ν - k - 1)/(k + 1)
+        a = a_pre / z₀
+        asum_prev = asum
+        asum += a
+        
+        if abs(asum_prev - asum) < ϵ
+            break
+        end
+        
+        Δ_prod_fact *= -Δ / (k + 2)
+    end
+
+    res = exp(-z₀) * asum
+    return res
+end
+
+# series about origin, general ν
+# https://functions.wolfram.com/GammaBetaErf/ExpIntegralE/06/01/04/01/01/0003/
+function En_expand_origin(ν::Number, z::Number)
+    if isinteger(ν)
+        # go to special case for integer ν
+        return En_expand_origin(Int(ν), z)
+    end
+    gammaterm = En_safe_gamma_term(ν, z)
+    frac = 1
+    sumterm = frac / (1 - ν)
+    k, maxiter = 1, 100
+    ϵ = 10*eps(real(sumterm))
+    while k < maxiter
+        frac *= -z / k
+        prev = sumterm
+        sumterm += frac / (k + 1 - ν)
+        if abs(sumterm - prev) < ϵ
+            break
+        end
+        k += 1
+    end
+    
+    return gammaterm - sumterm
+end
+
+# series about the origin, special case for integer n
+# https://functions.wolfram.com/GammaBetaErf/ExpIntegralE/06/01/04/01/02/0005/
+function En_expand_origin(n::Integer, z::Number)
+    gammaterm = 1 # (-z)^(n-1) / (n-1)!
+    for i = 1:n-1 
+        gammaterm *= -z / i
+    end
+
+    gammaterm *= digamma(n) - log(z)
+    sumterm = float(n == 1 ? 0 : 1 / (1 - n))
+    frac = 1
+    k, maxiter = 1, 100
+    ϵ = 10*eps(real(sumterm))
+    while k < maxiter
+        frac *= -z / k
+        # skip term with zero denominator
+        if k != n-1
+            prev = sumterm
+            sumterm += frac / (k + 1 - n)
+            if abs(sumterm - prev) < ϵ
+                break
+            end
+        end
+        k += 1
+    end
+    return gammaterm - sumterm
+end
+
+# can find imaginary part of E_ν(x) for x on negative real axis analytically
+# https://functions.wolfram.com/GammaBetaErf/ExpIntegralE/04/05/01/0003/
+function En_imagbranchcut(ν::Number, z::Number)
+    a = real(z)
+    e1 = exp(π*imag(ν))
+    e2 = Complex(cospi(real(ν)), -sinpi(real(ν)))
+    impart = π * im * e1 * e2 * exp((ν-1)*log(complex(a)) - loggamma(ν))
+    return imag(impart) * im # get rid of any real error
+end
+
+const ORIGIN_EXPAND_THRESH = 3
+"""
+    En(ν, z)
+
+Compute the exponential integral of `z` with order `ν`.
+
+External links: [DLMF](https://dlmf.nist.gov/8.19)
+"""
+function En(ν::Number, z::Number, niter::Int=1000)
+    if abs(ν) > 50
+        throw(ArgumentError("Unsupported order |ν| > 50"))
+    end
+    ν, z = promote(ν, float(z))
+    if typeof(z) <: Real && z < 0
+        throw(DomainError(z, "En will only return a complex result if called with a complex argument"))
+    end
+
+    if z == 0.0
+        if real(ν) > 0
+            return 1 / (ν - 1)
+        else
+            return oftype(z, Inf)
+        end
+    end
+    if ν == 0
+        return exp(-z) / z
+    elseif ν == 1 && real(z) > 0 && (typeof(z) == Float64 || typeof(z) == Complex{Float64})
```

Comment:
`z isa Union{Float64,ComplexF64}` is more compact (and can be checked statically)